DIPs/1NNN-RA.md

```diff
@@ -0,0 +1,409 @@
+* Overriding `opEquals` must also require the user to override `toHash` accordingly: two objects that are equal, must have the same hash value.
+* `opCmp` reveals an outdated design and implementation. Its presence was historically required by built-in associative arrays, which used binary trees needing ordering. The current implementation of associative arrays uses hashtables that lift the requirement. In
+addition, not all objects can be meaningfully ordered, so the best approach is to make comparison opt-in. Ordering comparisons in other class-based languages are done by means of interfaces, e.g. [`Comparable<T>` (Java)](https://docs.oracle.com/javase/7/docs/api/java/lang/Comparable.html) or [`IComparable<T>` (C#)](https://msdn.microsoft.com/en-us/library/4d7sx9hd.aspx).
+* The static method `factory` is a global dependency sink because it allows creating an instance of any class in the application from a string containing its name. Currently there is no way for a class to opt out. This feature creates [code bloat](https://forum.dlang.org/post/mr6bl7$26f5$1@digitalmars.com) in the generated executable. At best, class factory registration should be opt-in because only a small number of classes (none in the standard library) require the feature.
+* The current approach doesn't give the user the chance to opt in or out of certain functions (behaviours). There can be cases where the imposed methods don't make sense for the class type: ex. not all abstractions are of comparable types.
+* Because of the hidden `__mutex` member, and the fact that the D programming language supports function attributes, the design of `Object` is susceptible to the Fragile Base Class [Problem](https://www.javaworld.com/article/2073649/why-extends-is-evil.html): this states that a small and seemingly unrelated change in the Base class can lead to bugs and breakages in the Derived classes.
+
+To provide a real example of the proble the following code compiles:
+```D
+class C { int a; this(int) @safe {} }
+
+void main()
+{
+    C c = new C(1);
+    C[] a = [c, c, c];
+    assert(a == [c, c, c]);
+}
+```
+while the next section of code fails to compile with the message "incompatible types for array comparison: `C[]` and `C[3]`":
+```D
+class C { int a; this(int) @safe {} }
+
+@safe void main()
+{
+    C c = new C(1);
+    C[] a = [c, c, c];
+    assert(a == [c, c, c]);
+}
+```
+It fails because the it calls the non-safe `Object.opEquals` method in a safe function. In fact, just comparing 2 classes with no user-defined opEquals - `assert (c == c)` - will issue an error in @safe code: "`@safe` function `D main` cannot call `@system` function `object.opEquals`".
```

Comment:
This is a Romanianism I often use too: repeated references to "it". English only tolerates short-range references to "it", and seldom twice in the same sentence. In addiotion, here the two uses of "it" refer to distinct things: first time it's about the compilation, second time it's about the function if I understand correctly.

Reply:
I've now fixed this and 2 other minor fixes from your comments in https://github.com/Robert-Aron293/DIPs/pull/1